### DISCUSSION TITLE
Driver confusion

### DISCUSSION THREAD

**Discussion by szmarci** [microsoft/devicescript#655] [Q&A] [answered]
Hi!

I am trying to write a driver for the MPU9250 accelerometer & gyroscope module based on the existing DA213 driver.
I adjusted the code based on some sample and the datasheet of the chip:

```ts
export class MPU9250Driver extends I2CDriver implements AccelerometerDriver {
  async readSample() {
      const data: Buffer = await this.readRegBuf(0x3B, 6)
      const x = convert(data[0], data[1])
      const y = convert(data[2], data[3])
      const z = convert(data[4], data[5])
      return [x,y,z]
  }
  
  subscribe(cb: (sample: Vector3D) => Promise<void>): void {
      const _this = this
      setInterval(async () => {
          const s = await _this.readSample()
          await cb(s as Vector3D)
      }, 50)
  }
}
```

```ts
import { startAccelerometer } from "@devicescript/drivers"

export class MPU9250 {
  async startAccelerometer() {
      const driver = new MPU9250Driver()
      await driver.init();
      const acc = await startAccelerometer(driver, {})
      return acc
  }
}
```
main.ts:
```ts
import { MPU9250 } from "./mpu9250";

const accelerometer = new MPU9250();
const instance = await accelerometer.startAccelerometer()

instance.reading.subscribe(v => {
    console.log("from accelerometer client", v)
})
```

When I run this code the subscribe method in `MPU9250Driver` class is called by `AccelerometerServer`. Apparently this supposed to seed the data here, which it does btw:
 
<img width="514" alt="image" src="https://github.com/microsoft/devicescript/assets/1974290/cd6525c4-14c4-4642-9fa9-ded12983ad07">

However the `console.log` in main.ts `subscribe` callback is not triggered until I add this:

```diff ts
import { Accelerometer } from "@devicescript/core";
import { MPU9250 } from "./mpu9250";

const accelerometer = new MPU9250();
+const instance = await accelerometer.startAccelerometer()

const a = new Accelerometer('accelerometer2');

instance.reading.subscribe(v => {
    console.log("from accelerometer client", v)
})
```

Then it starts to log, which is weird to me because this `const a` object just hangs in nowhere.
Also this is changed from `all clients bound` to `missing servers`:

<img width="403" alt="Screenshot 2023-11-05 at 17 47 50" src="https://github.com/microsoft/devicescript/assets/1974290/52249d3e-fe5f-4fdc-81c1-eb590b16e919">

Read the documentation about this client & server architecture, but it still does not click to me.
Can anyone of you explain it to me?

Thank you!

**Comment by pelikhan** [comment]
The `startAccelerometer` returns a client that is already bound to your server so you don't need the "const a .... " line. You can simply remove it.

As for testing on hardware on the pico, make sure to configure the I2C pins you plan to use (see https://microsoft.github.io/devicescript/developer/board-configuration#i2c ). You're close to get it to work.

**Reply by szmarci** [reply]
> The startAccelerometer returns a client that is already bound to your server so you don't need the "const a .... " line. You can simply remove it.

But if I remove that line, the subscription does not work anymore for some reason.
I made a [repository](https://github.com/szmarci/devicescript_test) for the code. I'm using the pico board with pins set up correctly, I'm just stuck with this. 
Sure I must be missing something trivial here...

**Reply by pelikhan** [reply]
It's a bug in the accelerometer driver. Without hardware, it should not be using i2c. Fix incoming later today.

**Reply by pelikhan** [reply]
Pushed a fix to handle simulated environments https://github.com/microsoft/devicescript/commit/7129d817c939db1d860c3a5676d6e1467ed5bb44

**Reply by szmarci** [reply]
Thanks for taking a look. 
With actual hardware it supposed to work? Client instance subscription still not logging for me.
Also strange if I open the services/accelerometer node after a couple of seconds the whole BP08 (accelerometer) panel disappears:

<img width="392" alt="image" src="https://github.com/microsoft/devicescript/assets/1974290/800d3036-4912-4883-8ea8-e654c07fe8e3">

If I rerun the script it shows again.

**Reply by pelikhan** [reply]
The simulation was broken... now the client binding bug appears in hardware and sim :)

There's currently a bug in the client role assignment that seems to occur when one instantiates the client in an class instance function. Surprisingly, move the startAccelerometer to a function fixes the issue:
```

export async function startMPU9250() {
  const driver = new MPU9250Driver()
  await driver.init();
  const acc = await startAccelerometer(driver, {})
  return acc
}
```

**Reply by pelikhan** [reply]
see https://github.com/szmarci/devicescript_test/pull/1

**Reply by szmarci** [reply]
Ah I see, `this` binding it is. 
It also works when you change the class method to an arrow function:

```ts
export class MPU9250 {
    startAccelerometer = async () => {
        const driver = new MPU9250Driver()
        await driver.init();
        const acc = await startAccelerometer(driver, {})
        return acc
    }
}
```

**Reply by mmoskal** [reply, marked as answer]
The issue should be now fixed in v2.15.20
